backpropagation/network/neural_network.py

```diff
@@ -92,23 +96,26 @@ def _collect_activation_function_arguments_and_values(
                 Activation function arguments and values for the given
                 input for each layer.
         """
-        activation_arguments = list()
-        activation_values = [x]
-
-        value = x
-        for weight in self.weights:
-            argument = weight @ np.insert(value, 0, 1)
+        def single_pass(previous_parameters, weight):
+            _, value = previous_parameters
+            argument = weight @ np.insert(value, 0, 1, axis=0)
             value = self.activation_function.calculate_value(argument)
-            activation_arguments.append(argument)
-            activation_values.append(value)
+            return argument, value
+
+        activation_parameters = list(accumulate(
+            [(None, x)] + self.weights, single_pass)
+        )
+        activation_arguments, activation_values = zip(*activation_parameters)
```

Comment:
For loop was more readable and more perfomant.

Reply:
Out of curiosity do you have any benchmarks regarding the speed difference between for loops and accumulate?

Reply:
Well, I meant unzipping done in line 108 and this 109-110 magic.

Reply:
Fixed in ec76e254beec8723a922814b2492ace1fa938ed9

The zip trick is described in the official documentation in the zip section: https://docs.python.org/3/library/functions.html#zip